eneyj/docs/SPECIFICATION

```diff
@@ -1,48 +1,31 @@
 == Specification of eneyj ==
 
-(A preceding x means this is not implemented yet.
-A preceding * means that this is an idea and is not implemented yet.)
+1) Z1 ZObject
 
-1) Z Objects
+1.1) Everything is a ZObject.
+     ZObjects are the only data structure.
+     Code and data are both represented as ZObjects.
+     ZObjects are also called eneyj objects.
 
-1.1) Everything is a Z Object.
-     Z Objects are the only data structure.
-     Code and data are both represented as Z Objects.
-     Z Objects are also called eneyj objects.
+1.2) ZObjects are immutable.
 
-1.2) Z Objects are immutable.
+1.3) ZObjects are sets of key-value pairs.
 
-1.3) Z Objects are sets of key-value pairs.
+1.4) Keys of key-value pairs of ZObjects are either global or local.
 
-1.4) Keys of key-value pairs of Z Objects are either global or local.
+1.5) Global keys have the form ZnKn with n natural numbers.
 
-1.5) Global keys concatenate a Z Id, the letter K, and a number.
+1.6) Local keys have the form Kn with n consecutive natural numbers.
 
-1.6) Local keys concatenate the letter K and a number in the decimal system,
-     starting with 1 with no leading zeros.
+1.7) Every key on a ZObject must be unique.
 
-1.7) The number in local keys must be consecutive.
+1.8) Values of key-value pairs of all ZObjects must be ZObjects.
```

Comment:
I don't see how this is possible; there must be primitive values somewhere, or you get a "turtles all the way down" scenario! Take something of type Z6 (string) - for example Z28. The value of its Z6K1 key has to be the bare string, not another Z object, or how do you ever end up with the value 'eneyj'?

Reply:
So I've been thinking about this some more. Maybe 'string' is really the only primitive type needed? And any case where a string could appear COULD be replaced by a Z6 ZObject instead (but not indefinitely, at some point there has to be a string). Code (in function implementations), positive integers (the values) and any other numeric types we might have can all be represented ultimately as strings or key-value pairs with string values. We may run into a need for something else down the road, but I think what I'd suggest here is to simply say "Values of key-value pairs of all ZObjects must be either ZObjects or strings; keys with a string value can also take a string-type ZObject as value." Something like that maybe?

Reply:
Hmm, maybe this should work in a specification for references here as well? That's a case where you need to distinguish between the values of "Z99" as a reference and the string "Z99", so it is necessary to encapsulate the "Z99" value into a string ZObject.

Reply:
The serialization displays a simple string, but it is to be interpreted as a String object. Yes, this does mean its turtles all the way down, but since this is lazy evaluated, I think that's OK.

But I see your point and will think about it further.